targets/feather-stm32f405.json

```diff
@@ -0,0 +1,13 @@
+{
+  "inherits": ["cortex-m4"],
+  "build-tags": ["feather_stm32f405", "stm32f405", "stm32f4", "stm32"],
+  "cflags": [
+    "-Qunused-arguments"
+  ],
+  "linkerscript": "targets/stm32f405.ld",
+  "extra-files": [
+    "src/device/stm32/stm32f405.s"
+  ],
+  "flash-method": "command",
+  "flash-command": "dfu-util -a 0 --dfuse-address 0x08000000 -D {hex}"
```

Comment:
Please add the following

```
  "openocd-interface": "stlink-v2",
  "openocd-target": "stm32f4x"
```

If you have the following debugger, you will be able to flash with the following command.

```
$ tinygo flash --target feather-stm32f405 --size short --programmer jlink ./src/examples/blinky1
$ tinygo flash --target feather-stm32f405 --size short --programmer st-link-v2-1 ./src/examples/blinky1
$ tinygo flash --target feather-stm32f405 --size short --programmer cmsis-dap ./src/examples/blinky1
```

Reply:
sweet, I didn't know I could flash and debug via jlink straight from tinygo. I've been building + flashing + debugging as separate manual steps.

You do realize the feather does not have an STLink onboard, right? So would I still use that same `stlink-v2` value for the `openocd-interface` key in the json? 

And btw, for my new-ish version of OpenOCD, it ships with a bare `stlink` interface script that can detect the target interface version automatically, avoiding the need to specify, e.g., `-v2`

Reply:
Yes, I know that stlink is not on feather-stm32f405.
And I know that since 2016 or so, I've been using interface/stlink.cfg.

Maybe you can set the value to anything, so let's set it to jlink.

```
"openocd-interface": "jlink",
```

Reply:
I'm getting errors when trying to flash using this method, and I'm not very familiar with OpenOCD, maybe you can spot what I'm doing wrong? I'll continue researching the issue in the mean time.

```
andrew@trip /usr/local/src/tinygo/dev (feature/feather-stm32f405-i2c *)
⧉  cat targets/feather-stm32f405.json
{
  "inherits": ["cortex-m4"],
  "build-tags": ["feather_stm32f405", "stm32f405", "stm32f4", "stm32"],
  "cflags": [
    "-Qunused-arguments"
  ],
  "linkerscript": "targets/stm32f405.ld",
  "extra-files": [
    "src/device/stm32/stm32f405.s"
  ],
  "openocd-interface": "jlink",
  "openocd-target": "stm32f4x"
}
```
 
And when flashing the board:
```
andrew@trip /usr/local/src/tinygo/dev (feature/feather-stm32f405-i2c *)
⧉  tinygo flash --target feather-stm32f405 --size short --programmer jlink ./src/examples/blinky1
   code    data     bss |   flash     ram
   2952       4    4144 |    2956    4148
Open On-Chip Debugger 0.10.0+dev-01365-g583a65644-dirty (2020-07-31-09:26)
Licensed under GNU GPL v2
For bug reports, read
http://openocd.org/doc/doxygen/bugs.html
Info : auto-selecting first available session transport "jtag". To override use 'transport select <transport>'.
Info : J-Link EDU Mini V1 compiled Jul 17 2020 16:25:21
Info : Hardware version: 1.00
Info : VTarget = 3.292 V
Info : clock speed 2000 kHz
Error: JTAG scan chain interrogation failed: all ones
Error: Check JTAG interface, timings, target power, etc.
Error: Trying to use configured scan chain anyway...
Error: stm32f4x.cpu: IR capture error; saw 0x0f not 0x01
Warn : Bypassing JTAG setup events due to errors
Error: Invalid ACK (7) in DAP response
Error: JTAG-DP STICKY ERROR
Error: Invalid ACK (7) in DAP response
Error: JTAG-DP STICKY ERROR
  -- *snip* repeats many times --
Error: Invalid ACK (7) in DAP response
Error: JTAG-DP STICKY ERROR
Error: Invalid ACK (7) in DAP response
Error: JTAG-DP STICKY ERROR
in procedure 'program'
** OpenOCD init failed **
shutdown command invoked

error: failed to flash /tmp/tinygo111238838/main.hex: exit status 1
```

Reply:
```
Info : auto-selecting first available session transport "jtag". To override use 'transport select <transport>'.
```

I think you need an extra `"openocd-transport": "swd"` in the target JSON file, to set the transport to SWD (instead of JTAG). Most Cortex-M chips use SWD.

Reply:
awesome -- adding `"openocd-transport": "swd"` worked great!

and right you are, had to solder to the SWD pins on this board, it don't believe it even has JTAG

going to update this with your all's suggestions

```
andrew@trip ~/Code/tinygo/dev (feature/feather-stm32f405-i2c *)
⧉  tinygo flash --target feather-stm32f405 --size short --programmer jlink ./src/examples/blinky1
   code    data     bss |   flash     ram
   2908       0    4144 |    2908    4144
Open On-Chip Debugger 0.10.0+dev-01365-g583a65644-dirty (2020-07-31-09:26)
Licensed under GNU GPL v2
For bug reports, read
http://openocd.org/doc/doxygen/bugs.html
swd
Info : J-Link EDU Mini V1 compiled Jul 17 2020 16:25:21
Info : Hardware version: 1.00
Info : VTarget = 3.288 V
Info : clock speed 2000 kHz
Info : SWD DPIDR 0x2ba01477
Info : stm32f4x.cpu: hardware has 6 breakpoints, 4 watchpoints
Info : starting gdb server for stm32f4x.cpu on 3333
Info : Listening on port 3333 for gdb connections
target halted due to debug-request, current mode: Thread
xPSR: 0x01000000 pc: 0x08000900 msp: 0x20001000
Info : Reduced speed from 8000 kHz to 4000 kHz (maximum).
** Programming Started **
Info : device id = 0x10076413
Info : flash size = 1024 kbytes
** Programming Finished **
** Resetting Target **
shutdown command invoked
```